src/amo/components/AddAddonToCollection/index.js

```diff
@@ -0,0 +1,301 @@
+import translate from 'core/i18n/translate';
+import log from 'core/logger';
+import Select from 'ui/components/Select';
+import Notice from 'ui/components/Notice';
+import type { AddonType } from 'core/types/addons';
+import type { ErrorHandlerType } from 'core/errorHandler';
+import type { I18nType } from 'core/types/i18n';
+import type { DispatchFunc } from 'core/types/redux';
+import type {
+  CollectionId, CollectionsState, CollectionType,
+} from 'amo/reducers/collections';
+import type { UsersStateType } from 'amo/reducers/users';
+import type { ElementEvent } from 'core/types/dom';
+
+import './styles.scss';
+
+
+type Props = {|
+  addon: AddonType | null,
+  className?: string,
+  dispatch: DispatchFunc,
+  errorHandler: ErrorHandlerType,
+  i18n: I18nType,
+  loadingAddonsInCollections: boolean,
+  loadingUserCollections: boolean,
+  siteUserId: number | null,
+  // These are all user collections that the current add-on is a part of.
+  addonInCollections: Array<CollectionType> | null,
+  // These are all collections belonging to the user.
+  userCollections: Array<CollectionType> | null,
+  _window: typeof window | Object,
+|};
+
+type OnSelectOptionType = () => void;
+
+type SelectData = {| options: Array<Object>, disabled: boolean |};
+
+export class AddAddonToCollectionBase extends React.Component<Props> {
+  optionSelectHandlers: { [key: string]: OnSelectOptionType };
+
+  static defaultProps = {
+    _window: typeof window !== 'undefined' ? window : {},
+  };
+
+  constructor(props: Props) {
+    super(props);
+    this.optionSelectHandlers = {};
+  }
+
+  componentDidMount() {
+    // This runs as componentDidMount() to only load data client side,
+    // not server side.
+    this.loadDataIfNeeded();
+  }
+
+  componentWillReceiveProps(nextProps: Props) {
+    this.loadDataIfNeeded(nextProps);
+  }
+
+  loadDataIfNeeded(nextProps?: Props) {
+    const allProps = { ...this.props, ...nextProps };
+    const {
+      dispatch,
+      errorHandler,
+      loadingUserCollections,
+      userCollections,
+      siteUserId,
+    } = allProps;
+
+    if (errorHandler.hasError()) {
+      // Abort loadng data so that the error can be rendered.
+      return;
+    }
+
+    if (siteUserId && !loadingUserCollections && !userCollections) {
+      dispatch(fetchUserCollections({
+        errorHandlerId: errorHandler.id, userId: siteUserId,
+      }));
+    }
+  }
+
+  onSelectOption = (event: ElementEvent<HTMLSelectElement>) => {
+    event.preventDefault();
+    const key = event.target.value;
+    const handleOption = this.optionSelectHandlers[key];
+    if (handleOption) {
+      handleOption();
+    } else {
+      log.warn(`No handler for option: "${key}"`);
+    }
+  }
+
+  addToCollection(collection: CollectionType) {
+    const { addon, errorHandler, dispatch, siteUserId } = this.props;
+    if (!addon) {
+      throw new Error(
+        'Cannot add to collection because no add-on has been loaded');
```

Comment:
No, I don't think a user would ever see this error. This select box gets rendered on the add-on detail page so there will always be an add-on. If not, a 404 will be rendered. An error is thrown here just to help in debugging in case a developer breaks something.